ops/testing.py

```diff
@@ -0,0 +1,372 @@
+
+    class TestCharm(charm_cls):
+        on = TestEvents()
+
+    # Note: jam 2020-03-01 This is so that errors in testing say MyCharm has no attribute foo,
+    # rather than TestCharm has no attribute foo.
+    TestCharm.__name__ = charm_cls.__name__
+
+    unit_name = meta.name + '/0'
+    harness = TestingHarness(unit_name)
+    the_model = model.Model(unit_name, meta, harness._get_backend())
+    the_framework = framework.Framework(":memory:", "no-disk-path", meta, the_model)
+    the_charm = TestCharm(the_framework, meta.name)
+    # noinspection PyProtectedMember
+    harness._register_charm(the_charm)
+    return harness
+
+
+# noinspection PyProtectedMember
+class TestingHarness:
+    """This class represents a way to build up the model that will drive a test suite.
+
+    The model that is created is from the viewpoint of the charm that you are testing.
+
+    :ivar charm: The instance of Charm that is backed by this testing harness.
+    :type charm: CharmBase
+    """
+
+    def __init__(self, unit_name):
+        """Create a testing harness that can drive a Model"""
+        self.unit_name = unit_name
+        self._backend = _TestingModelBackend(unit_name)
+        self._relation_id_counter = 0
+        self.charm = None
+
+    def _get_backend(self):
+        return self._backend
+
+    @property
+    def model(self):
+        return self.charm.model
+
+    def _register_charm(self, charm):
+        if self.charm is not None:
+            raise RuntimeError(
+                "registering charm {} while {} is already registered".format(
+                    charm, self.charm))
+        self.charm = charm
+
+    def _next_relation_id(self):
+        rel_id = self._relation_id_counter
+        self._relation_id_counter += 1
+        return rel_id
+
+    def add_relation(self, relation_name, remote_app, remote_app_data={}):
+        """Declare that there is a new relation between this app and `remote_app`
+
+        TODO: Once relation_created exists as a Juju hook, it should be triggered by this code.
+
+        :param relation_name: The relation on Charm that is being related to
+        :param remote_app_data: Optional data bag that the remote application is sending
+          If remote_app_data is not empty, this should trigger
+          ``charm.on[relation_name].relation_changed(app)``
+        :return: The relation_id created by this add_relation.
+        :rtype: int
+        """
+        rel_id = self._next_relation_id()
+        self._backend._relation_ids_map.setdefault(relation_name, []).append(rel_id)
+        self._backend._relation_names[rel_id] = relation_name
+        self._backend._relation_list_map[rel_id] = []
+        self._backend._relation_data[rel_id] = {
+            remote_app: remote_app_data,
+            self._backend.unit_name: {},
+            self._backend.app_name: {},
+        }
+        # Reload the relation_ids list
+        self.model.relations._invalidate(relation_name)
```

Comment:
This stuff seems a bit fragile.. but not a blocker. As long as the external behavior is the one we want we can always improve these details.